File: go/getting_started.go
package main

import (
	"fmt"
	"log"

	"github.com/HelixDB/helix-go"
)

var HelixClient *helix.Client

type CreateUserResponse struct {
	User struct {
		ID        string      `json:"id"`
		Name      string      `json:"name"`
		Age       int32       `json:"age"`
		Email     string      `json:"email"`
		CreatedAt interface{} `json:"created_at"`
		UpdatedAt interface{} `json:"updated_at"`
	} `json:"user"`
}

type CreatePostResponse struct {
	Post struct {
		ID        string      `json:"id"`
		Content   string      `json:"content"`
		CreatedAt interface{} `json:"created_at"`
		UpdatedAt interface{} `json:"updated_at"`
	} `json:"post"`
}

type GetUsersResponse struct {
	Users []struct {
		ID        string      `json:"id"`
		Name      string      `json:"name"`
		Age       int32       `json:"age"`
		Email     string      `json:"email"`
		CreatedAt interface{} `json:"created_at"`
		UpdatedAt interface{} `json:"updated_at"`
	} `json:"users"`
}

type GetPostsResponse struct {
	Posts []struct {
		ID        string      `json:"id"`
		Content   string      `json:"content"`
		CreatedAt interface{} `json:"created_at"`
		UpdatedAt interface{} `json:"updated_at"`
	} `json:"posts"`
}

func main() {

	// connect to client
	HelixClient = helix.NewClient("http://localhost:6969")

	fmt.Println("Creating 3 users...")

	// create alice
	aliceData := map[string]any{
		"name":  "Alice",
		"age":   int32(25),
		"email": "alice@example.com",
	}

	var aliceUser CreateUserResponse
	err := HelixClient.Query("createUser", helix.WithData(aliceData)).Scan(&aliceUser)
	if err != nil {
		log.Fatalf("Error creating Alice: %s", err)
	}
	fmt.Printf("Created Alice: %+v\n", aliceUser)

	// create bob
	bobData := map[string]any{
		"name":  "Bob",
		"age":   int32(30),
		"email": "bob@example.com",
	}

	var bobUser CreateUserResponse
	err = HelixClient.Query("createUser", helix.WithData(bobData)).Scan(&bobUser)
	if err != nil {
		log.Fatalf("Error creating Bob: %s", err)
	}
	fmt.Printf("Created Bob: %+v\n", bobUser)

	// create charlie
	charlieData := map[string]any{
		"name":  "Charlie",
		"age":   int32(28),
		"email": "charlie@example.com",
	}

	var charlieUser CreateUserResponse
	err = HelixClient.Query("createUser", helix.WithData(charlieData)).Scan(&charlieUser)
	if err != nil {
		log.Fatalf("Error creating Charlie: %s", err)
	}
	fmt.Printf("Created Charlie: %+v\n", charlieUser)

	aliceID := aliceUser.User.ID
	bobID := bobUser.User.ID
	charlieID := charlieUser.User.ID

	fmt.Println("Creating follow relationships...")

	// alice follows bob
	followData1 := map[string]any{
		"follower_id": aliceID,
		"followed_id": bobID,
	}
	var followResult1 map[string]any
	err = HelixClient.Query("createFollow", helix.WithData(followData1)).Scan(&followResult1)
	if err != nil {
		log.Fatalf("Error creating follow relationship: %s", err)
	}
	fmt.Printf("Alice follows Bob: %+v\n", followResult1)

	// bob follows charlie
	followData2 := map[string]any{
		"follower_id": bobID,
		"followed_id": charlieID,
	}
	var followResult2 map[string]any
	err = HelixClient.Query("createFollow", helix.WithData(followData2)).Scan(&followResult2)
	if err != nil {
		log.Fatalf("Error creating follow relationship: %s", err)
	}
	fmt.Printf("Bob follows Charlie: %+v\n", followResult2)

	// charlie follows alice
	followData3 := map[string]any{
		"follower_id": charlieID,
		"followed_id": aliceID,
	}
	var followResult3 map[string]any
	err = HelixClient.Query("createFollow", helix.WithData(followData3)).Scan(&followResult3)
	if err != nil {
		log.Fatalf("Error creating follow relationship: %s", err)
	}
	fmt.Printf("Charlie follows Alice: %+v\n", followResult3)

	fmt.Println("Creating posts...")

	// alice's post
	alicePostData := map[string]any{
		"user_id": aliceID,
		"content": "Hello world! My first post on HelixDB",
	}

	var alicePost CreatePostResponse
	err = HelixClient.Query("createPost", helix.WithData(alicePostData)).Scan(&alicePost)
	if err != nil {
		log.Fatalf("Error creating Alice's post: %s", err)
	}
	fmt.Printf("Created Alice's post: %+v\n", alicePost)

	// bob's post
	bobPostData := map[string]any{
		"user_id": bobID,
		"content": "Loving the graph database capabilities",
	}

	var bobPost CreatePostResponse
	err = HelixClient.Query("createPost", helix.WithData(bobPostData)).Scan(&bobPost)
	if err != nil {
		log.Fatalf("Error creating Bob's post: %s", err)
	}
	fmt.Printf("Created Bob's post: %+v\n", bobPost)

	// charlie's post
	charliePostData := map[string]any{
		"user_id": charlieID,
		"content": "Building cool social networks with Helix",
	}

	var charliePost CreatePostResponse
	err = HelixClient.Query("createPost", helix.WithData(charliePostData)).Scan(&charliePost)
	if err != nil {
		log.Fatalf("Error creating Charlie's post: %s", err)
	}
	fmt.Printf("Created Charlie's post: %+v\n", charliePost)

	fmt.Println("Creating post embeddings with dummy vectors...")

	// alice's post embedding
	aliceEmbeddingData := map[string]any{
		"post_id": alicePost.Post.ID,
		"vector":  []float64{0.1, 0.2, 0.3, 0.4, 0.5},
		"content": "Hello world! My first post on HelixDB",
	}
	var aliceEmbeddingResult map[string]any
	err = HelixClient.Query("createPostEmbedding", helix.WithData(aliceEmbeddingData)).Scan(&aliceEmbeddingResult)
	if err != nil {
		log.Fatalf("Error creating Alice's post embedding: %s", err)
	}
	fmt.Printf("Created Alice's embedding: %+v\n", aliceEmbeddingResult)

	// bob's post embedding
	bobEmbeddingData := map[string]any{
		"post_id": bobPost.Post.ID,
		"vector":  []float64{0.2, 0.3, 0.4, 0.5, 0.6},
		"content": "Loving the graph database capabilities",
	}
	var bobEmbeddingResult map[string]any
	err = HelixClient.Query("createPostEmbedding", helix.WithData(bobEmbeddingData)).Scan(&bobEmbeddingResult)
	if err != nil {
		log.Fatalf("Error creating Bob's post embedding: %s", err)
	}
	fmt.Printf("Created Bob's embedding: %+v\n", bobEmbeddingResult)

	// charlie's post embedding
	charlieEmbeddingData := map[string]any{
		"post_id": charliePost.Post.ID,
		"vector":  []float64{0.3, 0.4, 0.5, 0.6, 0.7},
		"content": "Building cool social networks with Helix",
	}
	var charlieEmbeddingResult map[string]any
	err = HelixClient.Query("createPostEmbedding", helix.WithData(charlieEmbeddingData)).Scan(&charlieEmbeddingResult)
	if err != nil {
		log.Fatalf("Error creating Charlie's post embedding: %s", err)
	}
	fmt.Printf("Created Charlie's embedding: %+v\n", charlieEmbeddingResult)

	fmt.Println("\nQuerying and displaying the data...")

	// get all users
	fmt.Println("All Users:")
	var usersResponse GetUsersResponse
	err = HelixClient.Query("getUsers").Scan(&usersResponse)
	if err != nil {
		log.Fatalf("Error getting users: %s", err)
	}
	fmt.Printf("%+v\n", usersResponse)

	// get all posts
	fmt.Println("\nAll Posts:")
	var postsResponse GetPostsResponse
	err = HelixClient.Query("getPosts").Scan(&postsResponse)
	if err != nil {
		log.Fatalf("Error getting posts: %s", err)
	}
	fmt.Printf("%+v\n", postsResponse)

	// get alice's posts
	fmt.Println("\nAlice's Posts:")
	var alicePostsResponse GetPostsResponse
	alicePostsData := map[string]any{"user_id": aliceID}
	err = HelixClient.Query("getPostsByUser", helix.WithData(alicePostsData)).Scan(&alicePostsResponse)
	if err != nil {
		log.Fatalf("Error getting Alice's posts: %s", err)
	}
	fmt.Printf("%+v\n", alicePostsResponse)

	// get who alice is following
	fmt.Println("\nWho Alice is Following:")
	var aliceFollowingResponse map[string]any
	aliceFollowingData := map[string]any{"user_id": aliceID}
	err = HelixClient.Query("getFollowing", helix.WithData(aliceFollowingData)).Scan(&aliceFollowingResponse)
	if err != nil {
		log.Fatalf("Error getting Alice's following: %s", err)
	}
	fmt.Printf("%+v\n", aliceFollowingResponse)

	// get bob's followers
	fmt.Println("\nBob's Followers:")
	var bobFollowersResponse map[string]any
	bobFollowersData := map[string]any{"user_id": bobID}
	err = HelixClient.Query("getFollowers", helix.WithData(bobFollowersData)).Scan(&bobFollowersResponse)
	if err != nil {
		log.Fatalf("Error getting Bob's followers: %s", err)
	}
	fmt.Printf("%+v\n", bobFollowersResponse)

	// search for posts similar to alice's embedding
	fmt.Println("\nSearching for posts similar to Alice's embedding [0.1, 0.2, 0.3, 0.4, 0.5]:")
	var searchResponse GetPostsResponse
	searchData := map[string]any{
		"vector": []float64{0.1, 0.2, 0.3, 0.4, 0.5},
		"k":      int64(1),
	}
	err = HelixClient.Query("searchPostEmbeddings", helix.WithData(searchData)).Scan(&searchResponse)
	if err != nil {
		log.Fatalf("Error searching post embeddings: %s", err)
	}
	fmt.Printf("%+v\n", searchResponse)
}

File: typescript/src/getting_started.ts
import HelixDB from "helix-ts";

// type definitions for our data structures
interface User {
    id: string;
    name: string;
    age: number;
    email: string;
    created_at: string;
    updated_at: string;
}

interface Post {
    id: string;
    content: string;
    created_at: string;
    updated_at: string;
}

async function main() {
    // connect to client
    const client = new HelixDB("http://localhost:6969");

    console.log("Creating 3 users...");

    // create alice
    const aliceResponse = await client.query("createUser", {
        name: "Alice",
        age: 25,
        email: "alice@example.com"
    });
    console.log("Created Alice:", aliceResponse);

    // create bob
    const bobResponse = await client.query("createUser", {
        name: "Bob",
        age: 30,
        email: "bob@example.com"
    });
    console.log("Created Bob:", bobResponse);

    // create charlie
    const charlieResponse = await client.query("createUser", {
        name: "Charlie",
        age: 28,
        email: "charlie@example.com"
    });
    console.log("Created Charlie:", charlieResponse);

    const aliceId = aliceResponse.user.id;
    const bobId = bobResponse.user.id;
    const charlieId = charlieResponse.user.id;

    console.log("Creating follow relationships...");

    // alice follows bob
    const followResult1 = await client.query("createFollow", {
        follower_id: aliceId,
        followed_id: bobId
    });
    console.log("Alice follows Bob:", followResult1);

    // bob follows charlie
    const followResult2 = await client.query("createFollow", {
        follower_id: bobId,
        followed_id: charlieId
    });
    console.log("Bob follows Charlie:", followResult2);

    // charlie follows alice
    const followResult3 = await client.query("createFollow", {
        follower_id: charlieId,
        followed_id: aliceId
    });
    console.log("Charlie follows Alice:", followResult3);

    console.log("Creating posts...");

    // alice's post
    const alicePostResponse = await client.query("createPost", {
        user_id: aliceId,
        content: "Hello world! My first post on HelixDB"
    });
    console.log("Created Alice's post:", alicePostResponse);

    // bob's post
    const bobPostResponse = await client.query("createPost", {
        user_id: bobId,
        content: "Loving the graph database capabilities"
    });
    console.log("Created Bob's post:", bobPostResponse);

    // charlie's post
    const charliePostResponse = await client.query("createPost", {
        user_id: charlieId,
        content: "Building cool social networks with Helix"
    });
    console.log("Created Charlie's post:", charliePostResponse);

    console.log("Creating post embeddings with dummy vectors...");

    // alice's post embedding
    const aliceEmbeddingResult = await client.query("createPostEmbedding", {
        post_id: alicePostResponse.post.id,
        vector: [0.1, 0.2, 0.3, 0.4, 0.5],
        content: "Hello world! My first post on HelixDB"
    });
    console.log("Created Alice's embedding:", aliceEmbeddingResult);

    // bob's post embedding
    const bobEmbeddingResult = await client.query("createPostEmbedding", {
        post_id: bobPostResponse.post.id,
        vector: [0.2, 0.3, 0.4, 0.5, 0.6],
        content: "Loving the graph database capabilities"
    });
    console.log("Created Bob's embedding:", bobEmbeddingResult);

    // charlie's post embedding
    const charlieEmbeddingResult = await client.query("createPostEmbedding", {
        post_id: charliePostResponse.post.id,
        vector: [0.3, 0.4, 0.5, 0.6, 0.7],
        content: "Building cool social networks with Helix"
    });
    console.log("Created Charlie's embedding:", charlieEmbeddingResult);

    console.log("\nQuerying and displaying the data...");

    // get all users
    console.log("All Users:");
    const usersResponse = await client.query("getUsers", {});
    console.log(JSON.stringify(usersResponse, null, 2));

    // get all posts
    console.log("\nAll Posts:");
    const postsResponse = await client.query("getPosts", {});
    console.log(JSON.stringify(postsResponse, null, 2));

    // get alice's posts
    console.log("\nAlice's Posts:");
    const alicePostsResponse = await client.query("getPostsByUser", {
        user_id: aliceId
    });
    console.log(JSON.stringify(alicePostsResponse, null, 2));

    // get who alice is following
    console.log("\nWho Alice is Following:");
    const aliceFollowingResponse = await client.query("getFollowing", {
        user_id: aliceId
    });
    console.log(JSON.stringify(aliceFollowingResponse, null, 2));

    // get bob's followers
    console.log("\nBob's Followers:");
    const bobFollowersResponse = await client.query("getFollowers", {
        user_id: bobId
    });
    console.log(JSON.stringify(bobFollowersResponse, null, 2));

    // search for posts similar to alice's embedding
    console.log("\nSearching for posts similar to Alice's embedding [0.1, 0.2, 0.3, 0.4, 0.5]:");
    const searchResponse = await client.query("searchPostEmbeddings", {
        vector: [0.1, 0.2, 0.3, 0.4, 0.5],
        k: 1
    });
    console.log(JSON.stringify(searchResponse, null, 2));
}

main().catch(console.error);